python/lsst/meas/extensions/shapeHSM/_hsm_moments.py

```diff
@@ -0,0 +1,208 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import galsim
+import lsst.afw.geom as afwGeom
+import lsst.afw.table as afwTable
+import lsst.meas.base as measBase
+import lsst.pex.config as pexConfig
+from lsst.meas.base import (
+    FlagDefinitionList,
+    FlagHandler,
+    MeasurementError,
+    SingleFramePlugin,
+    SingleFramePluginConfig,
+)
+from lsst.pex.exceptions import InvalidParameterError
+
+
+class HsmMomentsConfig(SingleFramePluginConfig):
+    roundMoments = pexConfig.Field[bool](doc="Use round weight function?", default=False)
+    addFlux = pexConfig.Field[bool](doc="Store measured flux?", default=False)
+    subtractCenter = pexConfig.Field[bool](doc="Subtract starting center from x/y outputs?", default=False)
+
+
+class HsmMomentsPlugin(SingleFramePlugin):
+    ConfigClass = HsmMomentsConfig
+
+    def __init__(self, config, name, schema, logName=None):
+        super().__init__(config, name, schema, logName)
+        flagDefs = FlagDefinitionList()
+        self.FAILURE = flagDefs.add("FAILURE", "General failure flag, set if anything went wrong.")
+        # ... same with others
+        self.flagHandler = FlagHandler.addFields(schema, name, flagDefs)
+
+    @classmethod
+    def getExecutionOrder(cls):
+        return cls.SHAPE_ORDER
+
+    def fail(self, record, error=None):
+        # docstring inherited.
+        self.flagHandler.handleFailure(record)
+        if error:
+            center = record.getCentroid()
+            self.log.debug(
+                "Failed to measure shape for %d at (%f, %f): %s",
+                record.getId(),
+                center.getX(),
+                center.getY(),
+                error,
+            )
+
+
+class HsmSourceMomentsConfig(HsmMomentsConfig):
+    # later we can put some doscstring
+    badMaskPlanes = pexConfig.ListField[str](
+        doc="Mask planes used to reject bad pixels.", default=["BAD", "SAT"]
+    )
+
+
+@measBase.register("ext_shapeHSM_HsmSourceMoments")
+class HsmSourceMomentsPlugin(HsmMomentsPlugin):
+    ConfigClass = HsmSourceMomentsConfig
+
+    def __init__(self, config, name, schema, metadata, logName=None) -> None:
+        super().__init__(config, name, schema, logName)
+        # some flag thingy
+        # work in conjunction with fail()
+        self.centroidResultKey = afwTable.Point2DKey.addFields(
+            schema, "centroid", "The centroid measured by HSM", "pixel"
+        )
+        self.shapeKey = afwTable.QuadrupoleKey.addFields(schema, name, doc="HSM source moments.")
+        if config.addFlux:  # search the correct attribute name in the codebase
+            self.fluxKey = measBase.FluxResultKey.addFields(schema, "flux", "The flux measured by HSM", "dn")
+
+    def measure(self, record, exposure):
+        """... in place"""
+        # source in c++ is record in py. exposure the same
+
+        center = record.getCentroid()
```

Comment:
We should switch to getting the centroid safely. See:
https://github.com/lsst/meas_extensions_convolved/blob/42fff21dfb88e9ac1772aa792458ec200baa4b58/python/lsst/meas/extensions/convolved/convolved.py#L278
and
https://github.com/lsst/meas_extensions_convolved/blob/42fff21dfb88e9ac1772aa792458ec200baa4b58/python/lsst/meas/extensions/convolved/convolved.py#L311